README.rst
====================================
Conventional Commit Sphinx Changelog
====================================


.. image:: https://img.shields.io/pypi/v/cc_sphinx_changelog.svg
        :target: https://pypi.python.org/pypi/cc_sphinx_changelog

.. image:: https://github.com/pgierz/cc_sphinx_changelog/workflows/Python%20package/badge.svg
        :target: https://github.com/pgierz/cc_sphinx_changelog/actions?query=workflow%3A%22Python+package%22

.. image:: https://readthedocs.org/projects/convention-commits-sphinx/badge/?version=latest
        :target: https://convention-commits-sphinx.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status




Automatically generates a releases-style changelog based on conventional commit messages.


* Free software: GNU General Public License v3
* Documentation: https://convention-commits-sphinx.readthedocs.io.


Features
--------

Add the following to your Sphinx ReST files to get an automatically generated changelog based upon conventional commit messages::

    .. cc_changelog::


For more info on conventional commits, please see:

* https://www.conventionalcommits.org/en/v1.0.0/
* https://commitizen-tools.github.io/commitizen/

Here is an example of a changelog in ``git log --graph --oneline`` form::

    * 37b88f7 - (HEAD -> master, origin/master) docs(README): uses rtd link from project home (copy/paste) (64 seconds ago) <Paul Gierz>
    * 2106914 - docs(readme): fixes double dot in badge link (oops) (5 minutes ago) <Paul Gierz>
    * dba6788 - docs(README): fixes readme links (6 minutes ago) <Paul Gierz>
    * 87e04fc - (tag: v0.6.1) bump: version 0.6.0 → 0.6.1 (19 minutes ago) <GitHub Action>
    * afc76c2 - fix(setup.py): adds missing dependencies (20 minutes ago) <Paul Gierz>
    * 3a5de72 - test(tox): removes tox test suite (20 minutes ago) <Paul Gierz>
    * fa9923e - (tag: v0.6.0) bump: version 0.5.2 → 0.6.0 (26 minutes ago) <GitHub Action>
    * 17137dc - feat: adds sphinx directive (27 minutes ago) <Paul Gierz>
    * 1253061 - feat: adds optional printing of colors (default on) (28 minutes ago) <Paul Gierz>
    * 60d2aa6 - (tag: v0.5.2) bump: version 0.5.1 → 0.5.2 (2 days ago) <GitHub Action>
    * f025602 - refactor: moves main program into a simple function (2 days ago) <Paul Gierz>
    * acf0e3f - style: applies black and isort (2 days ago) <Paul Gierz>
    * 944e0b7 - docs: changes to rtd theme, adds heading to changelog (3 days ago) <Paul Gierz>
    * 89ad1e5 - docs: adds formatted changelog to docs (3 days ago) <Paul Gierz>
    * a09c9b0 - (tag: v0.5.1) bump: version 0.5.0 → 0.5.1 (3 days ago) <GitHub Action>
    * 55dec48 - fix: fixes heading grouping logic (3 days ago) <Paul Gierz>
    * 43a8501 - fix(setup.py): adds missing dependencies (3 days ago) <Paul Gierz>
    * 8997fa7 - (tag: v0.5.0) bump: version 0.4.1 → 0.5.0 (3 days ago) <Paul Gierz>
    *   dff4e2f - Merge remote-tracking branch 'origin/master' (3 days ago) <Paul Gierz>
    |\
    | * d25017d - (tag: v0.4.1) bump: version 0.4.0 → 0.4.1 (4 days ago) <GitHub Action>
    * | fb11339 - feat: adds color output and formats to stdout nicely (3 days ago) <Paul Gierz>
    * | 0dc1750 - (origin/feature/funcs) refactor: puts everything into functions (3 days ago) <Paul Gierz>
    |/
    * 9ea4503 - ci: automatically publish to pypi when automatic bump occurs (4 days ago) <Paul Gierz>
    * 06d4115 - fix: better implementation of getting commit messages (4 days ago) <Paul Gierz>
    * 0d9a1e5 - (tag: v0.4.0) bump: version 0.3.0 → 0.4.0 (4 days ago) <GitHub Action>
    * c9dd3c2 - feat: parses log and prints rST style release version summaries (4 days ago) <Paul Gierz>
    * df12a50 - docs(changelog): changelog fixup (4 days ago) <Paul Gierz>
    * 04209f6 - (tag: v0.3.0) bump: version 0.2.0 → 0.3.0 (4 days ago) <Paul Gierz>
    * 75a190a - feat: automatic bump version on push (4 days ago) <Paul Gierz>
    * c39e0fb - (tag: v0.2.0) bump: version 0.1.1 → 0.2.0 (4 days ago) <Paul Gierz>
    * 72f506a - fix(tags): adds v prefix to tags (4 days ago) <Paul Gierz>
    * 5724fb1 - ci: fixs tag publishing (4 days ago) <Paul Gierz>
    * fbd165c - ci: removes travis ci (4 days ago) <Paul Gierz>
    * 8b73303 - (tag: 0.1.1) bump: version 0.1.0 → 0.1.1 (4 days ago) <Paul Gierz>
    * 839fb8d - fix(cz): fixup conventional commit bumping (4 days ago) <Paul Gierz>
    * 2fc0466 - (tag: 0.1.0) bump: version 0.0.1 → 0.1.0 (4 days ago) <Paul Gierz>
    * f78d28e - feat: initial commit (4 days ago) <Paul Gierz>

Roadmap
-------

In the future, I'd like to include better formatting in Sphinx (preserve
paragraphs), as well as automatically create links to issue numbers and
releases.


CHANGELOG.md
## v0.6.1 (2020-12-28)

### Fix

- **setup.py**: adds missing dependencies

## v0.6.0 (2020-12-28)

### Feat

- adds sphinx directive
- adds optional printing of colors (default on)

## v0.5.2 (2020-12-26)

### Refactor

- moves main program into a simple function

## v0.5.1 (2020-12-25)

### Fix

- fixes heading grouping logic
- **setup.py**: adds missing dependencies

## v0.5.0 (2020-12-25)

### Feat

- adds color output and formats to stdout nicely

### Refactor

- puts everything into functions

## v0.4.1 (2020-12-24)

### Fix

- better implementation of getting commit messages

## v0.4.0 (2020-12-24)

### Feat

- parses log and prints rST style release version summaries

## v0.3.0 (2020-12-24)

### Feat

- automatic bump version on push

## v0.2.0 (2020-12-24)

### Fix

- **tags**: adds v prefix to tags

## 0.1.1 (2020-12-24)

### Fix

- **cz**: fixup conventional commit bumping

## 0.1.0 (2020-12-24)

### Feat

- initial commit


tests/__init__.py
# -*- coding: utf-8 -*-

"""Unit test package for cc_sphinx_changelog."""


tests/test_cc_sphinx_changelog.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `cc_sphinx_changelog` package."""


import unittest

from cc_sphinx_changelog import cc_sphinx_changelog


class TestCc_sphinx_changelog(unittest.TestCase):
    """Tests for `cc_sphinx_changelog` package."""

    def setUp(self):
        """Set up test fixtures, if any."""

    def tearDown(self):
        """Tear down test fixtures, if any."""

    def test_000_something(self):
        """Test something."""


cc_sphinx_changelog/__init__.py
# -*- coding: utf-8 -*-

"""Top-level package for Conventional Commit Sphinx Changelog."""

__author__ = """Paul Gierz"""
__email__ = "pgierz@awi.de"
__version__ = "0.6.1"


from docutils import nodes
from docutils.parsers.rst import Directive
from docutils.statemachine import ViewList
from sphinx.util.nodes import nested_parse_with_titles

from . import cc_sphinx_changelog


def make_section():
    section = nodes.section()
    section.attributes['ids'] = ["changelog"]
    section.attributes['names'] = ["changelog"]
    return section


class ConventionalCommitChangelog(Directive):

    def run(self):
        changelog = make_section()
        changelog += nodes.title(text="Changelog")

        repo = cc_sphinx_changelog.get_repo()
        commit_messages = cc_sphinx_changelog.get_commit_messages(repo)
        releases = cc_sphinx_changelog.group_commits(commit_messages, add_colors=False)


        for release in releases:
            release_section = make_section()
            release_section += nodes.title(text=release)
            changelog += release_section
            for heading, items in cc_sphinx_changelog.fmt_headings(releases[release].items()).items():
                if items:
                    heading_section = make_section()
                    heading_section.document = self.state.document
                    heading_section += nodes.title(text=heading)
                    release_section += heading_section
                    lst = nodes.bullet_list()
                    heading_section += lst
                    for indx, item in enumerate(items):
                        rst = ViewList()
                        cl_item = nodes.list_item()
                        lst += cl_item
                        rst.append(item, "changelog_tmp.rst", indx)
                        nested_parse_with_titles(self.state, rst, cl_item)
        return [changelog]


def setup(app):
    """Sphinx directive for making a changelog.

    Use the following to insert a formatted changelog:

    .. cc_changelog
    """
    app.add_directive("cc_changelog", ConventionalCommitChangelog)
    return {
        "version": __version__,
        "parallel_read_safe": True,
        "parallel_write_safe": True,
    }



cc_sphinx_changelog/cc_sphinx_changelog.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""Main module."""

import sys
import textwrap

import backpedal
import crayons
import git
from ansiwrap import ansilen


def ansi_ljust(s, width):
    needed = width - ansilen(s)
    if needed > 0:
        return s + " " * needed
    else:
        return s


def get_repo(repo_top=None):
    # TODO: Path from config (or URL??)
    if not repo_top:
        repo_top = backpedal.find(".git", item_type="directory")
    return git.Repo(repo_top)


def get_commit_messages(repo):
    log = repo.head.log()
    commit_messages = [
        (commit.hexsha[:7], commit.message.strip()) for commit in repo.iter_commits()
    ]
    return commit_messages


def group_commits(commit_messages, add_colors):
    releases = {"Unreleased": {}}
    current_release = releases["Unreleased"]
    for message in commit_messages:
        if message[1].startswith("bump:"):
            rel_head = message[1].split(" → ")[-1]
            releases[rel_head] = {}
            current_release = releases[rel_head]
        fmt_message(message, current_release, add_colors=add_colors)
    return releases


def fmt_message(message, current_release, add_colors=True):
    sha, message = message
    if message.startswith("bump:"):
        return
    message = message.split(":")
    commit_type, commit_message = message[0], ":".join(message[1:])
    if "\n" in commit_message:
        commit_header = commit_message.splitlines()[0].lstrip()
        commit_body = textwrap.fill(" ".join(commit_message.splitlines()[1:]).strip())
        commit_body = "\n".join(27 * " " + l for l in commit_body.splitlines())
        commit_message = f"{commit_header}\n\n{commit_body}"
    else:
        commit_message = commit_message.strip()
    scope = None
    if "(" in commit_type and ")" in commit_type:
        commit_type, scope = commit_type.split("(")
        scope = scope.replace(")", "")
    commits = current_release.setdefault(commit_type, [])
    if scope:
        if add_colors:
            prejust = f"{crayons.yellow(sha)} [{crayons.red(scope)}]"
        else:
            prejust = f"{sha} [{scope}]"
        fmted_message = f"{ansi_ljust(prejust, 25)}" + commit_message
    else:
        if add_colors:
            fmted_message = f"{ansi_ljust(crayons.yellow(sha), 25)}{commit_message}"
        else:
            fmted_message = f"{ansi_ljust(sha, 25)}{commit_message}"
    commits.append(fmted_message)


def fmt_headings(items):
    TYPES = {
        "fix": "Bug Fix",
        "feat": "New Feature",
        "ci": "Support",
        "docs": "Support",
        "perf": "Support",
        "refactor": "Support",
    }
    SHOWN_HEADINGS = {"Bug Fix": [], "New Feature": [], "Support": [], "Unknown": []}
    for heading, commits in items:
        group = SHOWN_HEADINGS.get(TYPES.get(heading, "Unknown"))
        for commit in commits:
            group.append(commit)
    for group in SHOWN_HEADINGS:
        if group == "Bug Fix":
            if len(SHOWN_HEADINGS[group]) > 1:
                SHOWN_HEADINGS["Bug Fixes"] = SHOWN_HEADINGS[group]
                del SHOWN_HEADINGS[group]
        elif group == "New Feature":
            if len(SHOWN_HEADINGS[group]) > 1:
                SHOWN_HEADINGS["New Features"] = SHOWN_HEADINGS[group]
                del SHOWN_HEADINGS[group]
    return SHOWN_HEADINGS


def print_changelog(location):
    if isinstance(location, str):
        location = open(location, "w")
    repo = get_repo()
    commit_messages = get_commit_messages(repo)
    releases = group_commits(commit_messages, add_colors=True)

    for release in releases:
        print(crayons.cyan(release), file=location)
        print(crayons.cyan("=" * len(release)), file=location)
        for heading, items in fmt_headings(releases[release].items()).items():
            if items:
                print(f"{crayons.green(heading)}", file=location)
                print(crayons.green("-" * len(heading)), file=location)
                for item in items:
                    print(f"* {item}", file=location)
        print("\n", file=location)
    location.close()


if __name__ == "__main__":
    print_changelog(sys.stdout)
